File: r/audit/api_ldc_join.py
# pattern: Imperative Shell
"""
Join API corpus (1987–1995) to LDC labeled parquet on normalized headline + pub_date.
Uses difflib for fuzzy matching (Levenshtein distance).
Then applies us_assign heuristic (desk/section) to compute ldc_heuristic_us.

Provenance: fuzzy matching adapted from R approach in 00_proc_and_matching_prep.R:456-491

Approved deviation: us_assign() is applied to matched API-side rows (the heuristic
faces API data pre-1986 in production, and LDC lacks keywords in parquet form).
Output ldc_heuristic_us is the heuristic's tri-state verdict from API-side descriptors.
"""

import re
from pathlib import Path
from datetime import date
from difflib import SequenceMatcher
import polars as pl

# Paths
API_CORPUS_DIR = Path("/Users/strophios/immigration_project/00_ML_data_expansion/00_explorer/api_corpus/")
LDC_LABELED = Path("/Users/strophios/immigration_project/00_ML_data_expansion/00_explorer/us_filter/ldc_labeled.parquet")
LDC_CORPUS_DIR = Path("/Users/strophios/immigration_project/00_ML_data_expansion/00_explorer/ldc_corpus/")
OUT_DIR = Path("/Users/strophios/immigration_project/00_ML_data_expansion/00_explorer/us_filter/audit")
OUT_PARQUET = OUT_DIR / "api_ldc_matched.parquet"
OUT_DIR.mkdir(parents=True, exist_ok=True)

def normalize_headline(x):
    """Normalize: lowercase, alphanumeric+space only."""
    if x is None:
        return ""
    return re.sub(r"[^a-z0-9 ]", "", x.lower())

def simple_us_assign(news_desk, section_name):
    """Simplified us_assign: desk/section heuristic.

    Returns:
        True if US, False if not-US, None if ambiguous/unknown
    """
    section_us = {"u.s.", "new york", "new york and region", "washington"}
    dsk_us = {
        "national desk",
        "metropolitan desk",
        "connecticut weekly desk",
        "westchester weekly desk",
        "long island weekly desk",
        "new york region",
        "new jersey weekly desk",
        "the city weekly desk"
    }

    section_non_us = {"world"}
    dsk_non_us = {"foreign desk"}

    meta_is_us = False
    meta_not_us = False

    if section_name:
        section_lower = section_name.lower()
        if section_lower in section_us:
            meta_is_us = True
        elif section_lower in section_non_us:
            meta_not_us = True

    if news_desk:
        desk_lower = news_desk.lower()
        if desk_lower in dsk_us:
            meta_is_us = True
        elif desk_lower in dsk_non_us:
            meta_not_us = True

    if meta_is_us and not meta_not_us:
        return True
    elif meta_not_us and not meta_is_us:
        return False
    else:
        return None

# Load LDC labeled
print("Loading LDC labeled...")
ldc_labeled = pl.read_parquet(LDC_LABELED)

# Per-year join + heuristic application
matched_all = []
join_summary = []

for year in range(1987, 1996):
    print(f"\n=== Year {year} ===")

    # Load API for this year
    api_path = API_CORPUS_DIR / f"{year}.parquet"
    if not api_path.exists():
        print(f"  API parquet not found, skipping")
        continue

    api_df = pl.read_parquet(api_path)
    api_count = len(api_df)
    print(f"  API rows: {api_count}")

    # Load LDC corpus for this year
    ldc_corpus_path = LDC_CORPUS_DIR / f"publication_year={year}"
    if not ldc_corpus_path.exists():
        print(f"  LDC corpus dir not found, skipping")
        continue

    # Read LDC corpus files (parquet partition)
    ldc_corpus = pl.read_parquet(ldc_corpus_path)

    # Join ldc_labeled with publication_date from corpus
    ldc_full = ldc_labeled.join(
        ldc_corpus.select(["id", "publication_date"]),
        on="id",
        how="inner"
    )

    ldc_count = len(ldc_full)
    print(f"  LDC rows: {ldc_count}")

    # Prepare for matching
    api = api_df.with_columns([
        pl.col("id").cast(pl.Utf8).alias("api_id"),
        pl.col("lead_paragraph").alias("api_lead"),
        (pl.col("pub_date").cast(pl.Utf8)).alias("pub_date_str"),
        pl.col("headline").map_elements(normalize_headline, return_dtype=pl.Utf8).alias("headline_norm")
    ]).select(["api_id", "api_lead", "pub_date_str", "headline_norm", "news_desk", "section_name"])

    ldc = ldc_full.with_columns([
        pl.col("id").cast(pl.Utf8).alias("ldc_id"),
        (pl.col("publication_date").cast(pl.Utf8)).alias("publication_date_str"),
        pl.col("headline").map_elements(normalize_headline, return_dtype=pl.Utf8).alias("headline_norm")
    ]).select(["ldc_id", "publication_date_str", "headline_norm", "stripped_text", "us_label"])

    # Convert to Python for faster processing
    api_rows = api.to_dicts()
    ldc_rows = ldc.to_dicts()

    # Group by date
    api_by_date = {}
    for row in api_rows:
        date_str = row["pub_date_str"]
        if date_str not in api_by_date:
            api_by_date[date_str] = []
        api_by_date[date_str].append(row)

    ldc_by_date = {}
    for row in ldc_rows:
        date_str = row["publication_date_str"]
        if date_str not in ldc_by_date:
            ldc_by_date[date_str] = []
        ldc_by_date[date_str].append(row)

    # Match by date
    year_matches = []
    date_count = 0
    unique_dates = sorted(set(ldc_by_date.keys()))

    for date_str in unique_dates:
        date_count += 1
        if date_count % 50 == 0:
            print(f"    Processing date {date_count}/{len(unique_dates)}")

        api_on_date = api_by_date.get(date_str, [])
        ldc_on_date = ldc_by_date.get(date_str, [])

        if not api_on_date or not ldc_on_date:
            continue

        # One-to-one matching: each LDC row matches at most one API row
        ldc_matched = set()

        for api_row in api_on_date:
            api_headline = api_row["headline_norm"]

            # Find best unmatched LDC match
            best_ldc_idx = None
            best_dist = float("inf")

            for ldc_idx, ldc_row in enumerate(ldc_on_date):
                if ldc_idx in ldc_matched:
                    continue

                ldc_headline = ldc_row["headline_norm"]

                # Use difflib SequenceMatcher for similarity ratio
                # Convert to distance (1 - ratio)
                ratio = SequenceMatcher(None, api_headline, ldc_headline).ratio()
                dist = 1 - ratio

                if dist < best_dist:
                    best_dist = dist
                    best_ldc_idx = ldc_idx

            # Match if dist <= 0.5 (equivalently, ratio >= 0.5)
            # This is roughly equivalent to Levenshtein distance <= 5 for medium-length headlines
            if best_ldc_idx is not None and best_dist <= 0.5:
                ldc_row = ldc_on_date[best_ldc_idx]
                api_row_ldc_orig = api_df.filter(pl.col("id") == api_row["api_id"].replace("nyt://article/", "")).to_dicts()

                # Compute heuristic
                heuristic_us = simple_us_assign(api_row.get("news_desk"), api_row.get("section_name"))

                year_matches.append({
                    "api_id": api_row["api_id"],
                    "ldc_id": ldc_row["ldc_id"],
                    "api_lead": api_row["api_lead"],
                    "ldc_stripped_text": ldc_row["stripped_text"],
                    "ldc_us_label": ldc_row["us_label"],
                    "ldc_heuristic_us": heuristic_us
                })
                ldc_matched.add(best_ldc_idx)

    if year_matches:
        matched_all.extend(year_matches)

        print(f"  Matched pairs: {len(year_matches)}")
        joinability_rate = len(year_matches) / ldc_count * 100 if ldc_count > 0 else 0
        print(f"  Joinability rate: {joinability_rate:.2f}%")

        join_summary.append({
            "year": year,
            "ldc_rows": ldc_count,
            "matched_pairs": len(year_matches),
            "joinability_rate": joinability_rate
        })
    else:
        print(f"  Matched pairs: 0")
        print(f"  Joinability rate: 0%")

# Write output
if matched_all:
    matched_df = pl.DataFrame(matched_all)
    print(f"\nWriting output to {OUT_PARQUET}...")
    matched_df.write_parquet(OUT_PARQUET)
    print(f"Output written: {len(matched_df)} rows")
else:
    print("\nNo matches found!")

# Print summary
print("\n=== Grand Summary (1987–1995) ===")
if join_summary:
    summary_df = pl.DataFrame(join_summary)
    print(summary_df)

    total_ldc = summary_df["ldc_rows"].sum()
    total_matched = summary_df["matched_pairs"].sum()
    print(f"\nTotal LDC rows (1987–1995): {total_ldc}")
    print(f"Total matched pairs: {total_matched}")
    if total_ldc > 0:
        print(f"Overall joinability rate: {total_matched / total_ldc * 100:.2f}%")

File: r/audit/api_ldc_join.R
# pattern: Imperative Shell
#
# Semantics-preserving completion strategy for us_assign() application.
# Deduplicates unique location strings and applies the pure-function location pattern logic.
#
# Sourced from nyt_location_checking.R lines 247-344 (us_assign function).
# OPTIMIZATION: Load each RDS file once and cache all needed data (locations + metadata).

suppressMessages({
  library(arrow)
  library(dplyr)
  library(stringr)
  library(tidyverse)
})

# ---- Load vendored us_assign heuristic ----
cat("=== Step 0: Load location gazetteer and constants ===\n")
source("r/vendored/us_assign.R")
cat(sprintf("Loaded location gazetteer (%d patterns), constants loaded\n\n", nrow(loc_df)))

# ---- Paths ----
OUT_PARQUET <- "/Users/strophios/immigration_project/00_ML_data_expansion/00_explorer/us_filter/audit/api_ldc_matched.parquet"
LDC_PARSED_DIR <- "/Users/strophios/immigration_project/00_ML_data_expansion/LDC2008T19/data/parsed_to_rds/"

# ---- Load matched parquet ----
cat("=== Step 1: Load matched parquet ===\n")
matched <- read_parquet(OUT_PARQUET) %>%
  mutate(ldc_heuristic_us = NA) %>%
  as_tibble()
cat(sprintf("Loaded %d matched pairs\n", nrow(matched)))

# ---- Step 2-4: Load RDS once, extract everything, build deduplicated cache ----
cat("\n=== Step 2-4: Load RDS, extract locations & metadata, deduplicate ===\n")

ldc_ids_set <- sort(unique(as.integer(matched$ldc_id)))
cat(sprintf("Need keywords for %d unique LDC IDs\n", length(ldc_ids_set)))

# Single pass through RDS files: extract locations, metadata, and build location->verdict cache
location_strings <- c()
article_locs_list <- list()  # will be converted to tibble after loop
article_ids_list <- list()
ldc_meta_list <- list()

for (year in 1987:1995) {
  rds_path <- file.path(LDC_PARSED_DIR, paste0(year, ".rds"))
  if (!file.exists(rds_path)) next

  cat(sprintf("  Loading year %d RDS (%.0f MB)...\n", year, file.size(rds_path) / 1e6))
  ldc_year_full <- readRDS(rds_path)

  ids_in_year <- as.integer(ldc_year_full$id)
  rows_to_keep <- which(ids_in_year %in% ldc_ids_set)

  if (length(rows_to_keep) == 0) {
    cat(sprintf("    No matched articles in year %d\n", year))
    next
  }

  cat(sprintf("    Processing %d matched articles...\n", length(rows_to_keep)))

  # Vectorized extraction: extract all locations and metadata at once
  locs_per_article <- map(rows_to_keep, function(idx) {
    descs <- ldc_year_full$descriptors[[idx]]
    if (is.data.frame(descs)) {
      loc_rows <- descs[str_detect(descs$type, "^location$"), , drop = FALSE]
      if (nrow(loc_rows) > 0) {
        return(loc_rows$content)
      }
    }
    return(character(0))
  })

  # Extract all location strings
  for (locs in locs_per_article) {
    location_strings <- c(location_strings, locs)
  }

  # Add to lists (much faster than add_row per row)
  article_ids_list[[length(article_ids_list) + 1]] <- as.character(ids_in_year[rows_to_keep])
  article_locs_list[[length(article_locs_list) + 1]] <- locs_per_article

  # Extract metadata
  ldc_meta_year <- tibble(
    ldc_id = as.character(ids_in_year[rows_to_keep]),
    section_name = ldc_year_full$online_sections[rows_to_keep],
    news_desk = ldc_year_full$dsk[rows_to_keep]
  )
  ldc_meta_list[[length(ldc_meta_list) + 1]] <- ldc_meta_year

  cat(sprintf("    Extracted: %d locations, %d metadata entries\n",
              length(location_strings), length(rows_to_keep)))
}

# Convert lists to tibbles
ldc_meta_full <- bind_rows(ldc_meta_list) %>% distinct(ldc_id, .keep_all = TRUE)
article_ids_vec <- unlist(article_ids_list)
article_locs_vec <- unlist(article_locs_list, recursive = FALSE)
article_locations <- tibble(ldc_id = article_ids_vec, location_list = article_locs_vec)

ldc_meta_full <- ldc_meta_full %>% distinct(ldc_id, .keep_all = TRUE)
unique_locations <- sort(unique(location_strings))

cat(sprintf("\nTotal unique location strings: %d\n", length(unique_locations)))
cat(sprintf("Total articles with location data: %d\n", nrow(article_locations)))

if (length(unique_locations) > 200000) {
  stop(sprintf("ERROR: Unique location count %d exceeds threshold. Aborting.", length(unique_locations)))
}

# ---- Step 5: Create synthetic dataframe and run us_assign ----
cat("\n=== Step 5: Run us_assign on unique locations ===\n")

synthetic_df <- tibble(
  id = seq_along(unique_locations),
  news_desk = NA_character_,
  section_name = NA_character_,
  keywords = map(unique_locations, function(loc_str) {
    data.frame(type = "location", value = loc_str, stringsAsFactors = FALSE)
  })
)

cat(sprintf("Synthetic df: %d unique location rows\n", nrow(synthetic_df)))
cat("Applying us_assign to unique locations...")
start_time <- Sys.time()
result_df <- us_assign(synthetic_df, return_long = FALSE)
elapsed <- difftime(Sys.time(), start_time, units = "secs")
cat(sprintf(" (%.1f seconds)\n", as.numeric(elapsed)))

location_verdicts <- tibble(
  location = unique_locations,
  loc_is_us = result_df$is_us,
  loc_not_us = result_df$not_us
)

cat("Verdict distribution (unique locations):\n")
print(location_verdicts %>%
      mutate(verdict = case_when(
        loc_is_us & !loc_not_us ~ "US",
        loc_not_us & !loc_is_us ~ "non-US",
        TRUE ~ "ambiguous"
      )) %>%
      count(verdict))

# ---- Step 6: Aggregate verdicts per article ----
cat("\n=== Step 6: Aggregate location verdicts to articles ===\n")

article_loc_verdicts <- article_locations %>%
  mutate(
    kw_is_us = map_lgl(location_list, function(locs) {
      if (length(locs) == 0) return(FALSE)
      any(location_verdicts[location_verdicts$location %in% locs, ]$loc_is_us, na.rm = TRUE)
    }),
    kw_not_us = map_lgl(location_list, function(locs) {
      if (length(locs) == 0) return(FALSE)
      any(location_verdicts[location_verdicts$location %in% locs, ]$loc_not_us, na.rm = TRUE)
    })
  ) %>%
  select(ldc_id, kw_is_us, kw_not_us)

cat(sprintf("Aggregated: %d articles\n", nrow(article_loc_verdicts)))

# ---- Step 7: Combine with metadata and create final verdict ----
cat("\n=== Step 7: Combine metadata and create tri-state verdict ===\n")

matched_with_verdicts <- matched %>%
  left_join(ldc_meta_full, by = "ldc_id") %>%
  left_join(article_loc_verdicts, by = "ldc_id") %>%
  mutate(
    # Meta flags from section/desk
    meta_is_us = str_to_lower(section_name) %in% section_us |
                 str_to_lower(news_desk) %in% dsk_us,
    meta_not_us = str_to_lower(section_name) %in% section_non_us |
                  str_to_lower(news_desk) %in% dsk_non_us
  )

# Handle multi-section names
rns <- which(str_detect(matched_with_verdicts[["section_name"]], "; "))
if (length(rns) > 0) {
  cat(sprintf("Processing %d multi-section articles\n", length(rns)))
  sec_rns_us <- map(rns, function(rn) {
    sec <- str_to_lower(matched_with_verdicts[["section_name"]][rn])
    sec <- str_split(sec, "; ")[[1]]
    sec_us <- any(sec %in% section_us)
    sec_not <- any(sec %in% section_non_us)
    tibble(meta_is_us = sec_us, meta_not_us = sec_not)
  }) %>% bind_rows()

  matched_with_verdicts[["meta_is_us"]][rns] <- sec_rns_us[["meta_is_us"]] |
    matched_with_verdicts[["meta_is_us"]][rns]
  matched_with_verdicts[["meta_not_us"]][rns] <- sec_rns_us[["meta_not_us"]] |
    matched_with_verdicts[["meta_not_us"]][rns]
}

# Create tri-state verdict
matched_with_verdicts <- matched_with_verdicts %>%
  mutate(
    any_us = kw_is_us | meta_is_us,
    any_not_us = kw_not_us | meta_not_us,
    ldc_heuristic_us = case_when(
      any_us & !any_not_us ~ TRUE,
      any_not_us & !any_us ~ FALSE,
      TRUE ~ NA
    )
  )

cat("\nTri-state verdict distribution:\n")
print(matched_with_verdicts %>% count(ldc_heuristic_us))

# ---- Step 8: Write output ====
cat("\n=== Step 8: Write output ===\n")

final_output <- matched_with_verdicts %>%
  select(api_id, ldc_id, api_lead, ldc_stripped_text, ldc_us_label, ldc_label_source, ldc_heuristic_us) %>%
  as.data.frame()

cat(sprintf("Final output: %d rows\n", nrow(final_output)))
write_parquet(final_output, OUT_PARQUET)

# Verify
cat("Verifying written parquet...\n")
verified <- read_parquet(OUT_PARQUET)
cat(sprintf("Verified: %d rows, %d columns\n", nrow(verified), ncol(verified)))

# ---- Final Report ----
cat("\n=== FINAL REPORT ===\n")
cat(sprintf("Unique location strings deduced: %d\n", length(unique_locations)))
cat(sprintf("us_assign time: %.1f seconds (3,155 patterns × %d strings)\n", as.numeric(elapsed), length(unique_locations)))
cat(sprintf("Matched pairs in output: %d (unchanged)\n", nrow(final_output)))
cat("\nTri-state verdict distribution:\n")
print(verified %>% as_tibble() %>% count(ldc_heuristic_us))

cat("\nComparison with dateline heuristic:\n")
comparison <- verified %>%
  as_tibble() %>%
  mutate(
    ldc_us_label = as.logical(ldc_us_label),
    disagreement = !is.na(ldc_heuristic_us) & !is.na(ldc_us_label) & (ldc_heuristic_us != ldc_us_label)
  ) %>%
  summarise(
    total_pairs = n(),
    both_non_null = sum(!is.na(ldc_heuristic_us) & !is.na(ldc_us_label)),
    na_heuristic = sum(is.na(ldc_heuristic_us)),
    na_dateline = sum(is.na(ldc_us_label)),
    disagreements = sum(disagreement, na.rm = TRUE),
    disagreement_rate = round(sum(disagreement, na.rm = TRUE) / sum(!is.na(ldc_heuristic_us) & !is.na(ldc_us_label)), 4)
  )
print(comparison)

cat("\n=== COMPLETION ===\n")
cat("Output written to:", OUT_PARQUET, "\n")
